### PR TITLE
[nnc] Update cuda codegen to use llvm for thread and block extent computations

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72077
* #72036
* #72038
* __->__ #72040
* #72032
* #71651
* #71650
* #71173

Differential Revision: [D33865041](https://our.internmc.facebook.com/intern/diff/D33865041)

### CODE REVIEW

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33865041).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33865041).

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 1]
```diff
@@ -1024,6 +1024,46 @@ void CudaCodeGen::Initialize() {
...
+    vars_in_extents.insert(v.begin(), v.end());
+  }
+  for (const auto& te : thread_extents) {
+    auto v = VarFinder::find(te);
+    vars_in_extents.insert(v.begin(), v.end());
+  }
+  for (const size_t i : c10::irange(buffer_args.size())) {
+    if (vars_in_extents.count(buffer_args[i].var())) {
+      extents_buffer_args.push_back(buffer_args[i]);
+      arg_pos_in_extents_.push_back(true);
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:1042]: Add `.reserve` for this.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 2]
```diff
@@ -1024,6 +1024,46 @@ void CudaCodeGen::Initialize() {
     thread_block_size_ = -1;
   }
 
+  // Build an LLVM based eval expression for the extents
+  block_extents_eval_.reserve(block_extents.size());
+  std::vector<BufferArg> extents_buffer_args;
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:1029]: Can you add the following comment here?

```
We need to extract the args that are used in the thread and block extents 
from bufferArgs and only use those for the `ExprEval` below. Without this, 
bufferArgs might contain arbitrary types that are not handled by LLVM and 
hence would result in an error.
```

**Review by navahgar — APPROVED** [review]
I thought you were going to use Mike's approach here. Something changed with that?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp:1042]: this is in initialize, not hot path, so perf doesn't matter as much here. but i can add

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33865041).

**Pull request closed without merging**